neuralmonkey/runners/beamsearch_runner.py

```diff
@@ -1,58 +1,147 @@
+# pylint: enable=unused-import
 from neuralmonkey.vocabulary import Vocabulary, END_TOKEN
 
+BeamSearchOutput = NamedTuple("BeamSearchOutput",
+                              [("step", int),
+                               ("scores", List[List[float]]),
+                               ("parent_ids", List[List[int]]),
+                               ("token_ids", List[List[int]])])
+
 
 class BeamSearchExecutable(Executable):
     def __init__(self,
                  rank: int,
                  all_encoders: Set[ModelPart],
-                 bs_outputs: SearchStepOutput,
+                 num_sessions: int,
+                 decoder: BeamSearchDecoder,
                  vocabulary: Vocabulary,
+                 beam_size: int,
                  postprocess: Optional[Callable]) -> None:
 
         self._rank = rank
+        self._num_sessions = num_sessions
         self._all_encoders = all_encoders
-        self._bs_outputs = bs_outputs
+        self._decoder = decoder
         self._vocabulary = vocabulary
         self._postprocess = postprocess
+        self._beam_size = beam_size
+        self._output = BeamSearchOutput(
+            step=0,
+            scores=np.empty([0, self._beam_size], dtype=float),
+            parent_ids=np.empty([0, self._beam_size], dtype=int),
+            token_ids=np.empty([0, self._beam_size], dtype=int))
+
+        self._next_feed = [{} for _ in range(self._num_sessions)] \
+            # type: List[FeedDict]
+
+        # If we are ensembling, we run only one step at a time
+        if self._num_sessions > 1:
+            for fd in self._next_feed:
+                fd.update({self._decoder.max_steps: 1})
 
         self.result = None  # type: ExecutionResult
 
     def next_to_execute(self) -> NextExecute:
-        return self._all_encoders, {'bs_outputs': self._bs_outputs}, {}
+        return (self._all_encoders,
+                {'bs_outputs': self._decoder.outputs},
+                self._next_feed)
 
+    # pylint: disable=too-many-locals
     def collect_results(self, results: List[Dict]) -> None:
-        if len(results) > 1:
-            raise ValueError("Beam search runner does not support ensembling.")
-
-        evaluated_bs = results[0]['bs_outputs']
-        max_time = evaluated_bs.scores.shape[0]
-
-        # pick the end of the hypothesis based on its rank
-        hyp_index = np.argpartition(
-            -evaluated_bs.scores[-1], self._rank - 1)[self._rank - 1]
-        bs_score = evaluated_bs.scores[-1][hyp_index]
-
-        # now backtrack
-        output_tokens = []  # type: List[str]
+        # Recompute logits
+        # Only necessary when ensembling models
+        prev_logprobs = []
+        for sess_idx, _ in enumerate(results):
+            bs_outputs = results[sess_idx]['bs_outputs']
+            prev_logprobs.append(bs_outputs.last_search_state.prev_logprobs)
+
+        prev_logprobs = np.divide(np.sum(prev_logprobs, 0), self._num_sessions)
```

Comment:
Takže při tom ensemblování používáme geometrický průměr? Nemělo by to být konfigurovatelne?